GAE/src/org/akvo/flow/rest/ImageUploadRestService.java

```diff
@@ -0,0 +1,219 @@
+        if (form == null) {
+            return new Response(400, "Form not found");
+        }
+        Question question = getQuestion(questionId);
+        if (question == null) {
+            return new Response(400, "Question not found");
+        }
+        if (question.getSurveyId() == null || !question.getSurveyId().equals(form.getObjectId())) {
+            return new Response(400, "Question does not belong to that form");
+        }
+        SurveyInstance formInstance = getFormInstance(formInstanceId);
+        if (formInstance == null) {
+            return new Response(400, "FormInstance not found");
+        }
+        if (formInstance.getSurveyId()== null || !formInstance.getSurveyId().equals(form.getObjectId())) {
+            return new Response(400, "FormInstance does not belong to that form");
+        }
+        if (file == null || file.isEmpty()) {
+            return new Response(400, "File is not valid");
+        }
+        String fileExtension = getFileType(file);
+        if (fileExtension == null) {
+            return new Response(400, "File type is not valid: only jpg and png are accepted");
+        }
+        //TODO: shall we resize? does the calling task handle resize
+        String filename = generateFileName(fileExtension);
+        String resultFilename = uploadImageToS3(file, filename);
+        if (resultFilename == null) {
+            return new Response(400, "Upload to s3 failed for: " + filename);
+        }
+        saveQuestionAnswer(question, formInstance, resultFilename);
+        return new Response(responseCode, errorMessage);
+    }
+
+    @Nullable
+    private String getFileType(MultipartFile file) {
+        //TODO: shall we accept other formats?
+        String contentType = file.getContentType() != null? file.getContentType(): "";
+        switch (contentType) {
+            case "image/jpeg":
+                return ".jpg";
+            case "image/png":
+                return ".png";
+        }
+        return null;
+    }
+
+    @Nullable
+    private String uploadImageToS3(MultipartFile file, String fileName) {
+        try {
+            Properties props = System.getProperties();
+            String bucketName = props.getProperty("s3bucket");
+            String directory = props.getProperty("images");
+            S3Util.put(bucketName, directory + "/" + fileName, file.getBytes(), file.getContentType(), true);
+            return fileName;
+        } catch (Exception e) {
+            log.log(Level.SEVERE, "Error uploading image", e);
+            return null;
+        }
+    }
+
+    private void saveQuestionAnswer(Question question, SurveyInstance formInstance, String fileName) {
+        Long questionGroupId = question.getQuestionGroupId();
+        QuestionGroup group = new QuestionGroupDao().getByKey(questionGroupId);
+        boolean repeatable = group.getRepeatable();
+        if (repeatable) {
+            saveRepeatedQuestionAnswer(question, formInstance, fileName);
```

Comment:
I think we should leave out the handling of repeat question group for now.  In this case, there are no photos in the repeated groups I believe. and its one that needs a little more designing on how to handle it.